react-components/src/components/NodeCacheProvider/PointCloudAnnotationCacheProvider.tsx

```diff
@@ -0,0 +1,179 @@
+
+import { type UseQueryResult, useQuery } from '@tanstack/react-query';
+import { useSDK } from '../RevealContainer/SDKProvider';
+import { useRevealKeepAlive } from '../RevealKeepAlive/RevealKeepAliveContext';
+import { PointCloudAnnotationCache } from './PointCloudAnnotationCache';
+import { type TypedReveal3DModel } from '../Reveal3DResources/types';
+import { type AnnotationAssetMappingDataResult } from '../../hooks/useClickedNode';
+import { type AnnotationModelDataResult } from '../../hooks/useCalculatePointCloudModelsStyling';
+
+export type PointCloudAnnotationCacheContextContent = {
+  cache: PointCloudAnnotationCache;
+};
+
+const PointCloudAnnotationCacheContext = createContext<
+  PointCloudAnnotationCacheContextContent | undefined
+>(undefined);
+
+const usePointCloudAnnotationCache = (): PointCloudAnnotationCache => {
+  const content = useContext(PointCloudAnnotationCacheContext);
+
+  if (content === undefined) {
+    throw Error('Must use usePointCloudAnnotationCache inside a PointCloudAnnotationCacheContext');
+  }
+
+  return content.cache;
+};
+
+export const usePointCloudAnnotationMappingsForModels = (
+  models: TypedReveal3DModel[]
+): UseQueryResult<AnnotationModelDataResult[]> => {
+  const pointCloudAnnotationCache = usePointCloudAnnotationCache();
+
+  return useQuery(
+    [
+      'reveal',
+      'react-components',
+      'models-pointcloud-annotations-mappings',
+      ...models.map((model) => `${model.modelId}/${model.revisionId}`).sort()
+    ],
+    async () => {
+      return await Promise.all(
+        models.map(async (model) => {
+          const annotationModel = await pointCloudAnnotationCache.getPointCloudAnnotationsForModel(
+            model.modelId,
+            model.revisionId
+          );
+          return {
+            model,
+            annotationModel
+          };
+        })
+      );
+    },
+    { staleTime: Infinity, enabled: models.length > 0 }
+  );
+};
+
+export const usePointCloudAnnotationMappingsForAssetIds = (
+  models: TypedReveal3DModel[],
+  assetIds: Array<string | number> | undefined
+): UseQueryResult<AnnotationAssetMappingDataResult[]> => {
+  const pointCloudAnnotationCache = usePointCloudAnnotationCache();
+
+  return useQuery(
+    [
+      'reveal',
+      'react-components',
+      'all-annotation-mappings',
+      ...models.map((model) => `${model.modelId}/${model.revisionId}`).sort(),
+      ...(assetIds?.map((assetId) => assetId.toString()).sort() ?? [])
+    ],
+    async () => {
+      return await Promise.all(
+        models.map(async (model) => {
+          await Promise.all(
+            assetIds?.map(async (assetId) => {
+              const result = await fetchAnnotationsForAssetId(
+                model.modelId,
+                model.revisionId,
+                assetId,
+                pointCloudAnnotationCache
+              );
+              return result ?? [];
+            }) ?? []
+          );
+        })
+      );
```

Comment:
This seems like an extremely network-heavy function, especially if there are many `assetIds`. What do we use it for? Can we send multiple asset IDs in just one call instead?